WFE returns headers Boulder-Request-ID and Boulder-Requester

benileo opened the pull request:
In this PR, two additional HTTP headers are added to aid in external error reporting and log analysis. 

1. Every response, apart from the path "/", contains a unique Boulder-Request-ID. 
2. Where present, the response will contain the registration ID in the header Boulder-Requester.

Fixes #1884

jsha commented on wfe/context.go:
```diff
@@ -65,6 +65,11 @@ func (th *topHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	defer th.logEvent(logEvent)
+	defer func() {
+		if logEvent.Requester > 0 {
+			w.Header().Set("Boulder-Requester", fmt.Sprintf("%d", logEvent.Requester))
```
I believe this will get called after `response.WriteHeader(code)`. We need to find another way to arrange for this to be called on every request that has a requester. One simple option that leads to a bit of duplication would be to set the header right after each `verifyPOST` call. This would be analagous to how we set the `logEvent.Requester` and `logEvent.Contacts` fields. Currently this is done inside `verifyPOST` (with one random exception in NewRegistration), but I would be okay with moving that to consistently happen at the call sites instead. This would be to avoid having to pass `response` into verifyPOST, which I would *really* not like to do.

Alternately we could refactor WFE so `verifyPOST` is called earlier in the handler stack, and the result passed along to lower values. This would give a natural single place to add the header, and would probably be a net improvement, but would be a very big refactoring.

benileo replied:
Since refactoring `verifyPOST` is not something I would be fully comfortable with at this point, I chose the former. It leads to a bit of duplication, but after each call to `verifyPOST` the Boulder-Requester header is added. Avoiding passing `response` into verifyPOST.

ccppuu commented on wfe/wfe_test.go:
```diff
@@ -1592,3 +1592,46 @@ func TestVerifyPOSTInvalidJWK(t *testing.T) {
+		URL:    mustParseURL(directoryPath),
+	})
+
+	test.AssertNotEquals(t, responseWriter.Header().Get("Boulder-Request-ID"), "")
```
It would be nice to see a test for the non-empty `Boulder-Request-ID` path as well. It seems like the primary challenge there is that the ID is randomly generated using `core.NewToken` and can't be predicted by the test unless we mock it or use a seeded RNG.

@jsha @benileo Do you think it's worth the trouble of figuring out how to mock out a predictable `core.NewToken` for testing or should we accept as-is with the empty value test?

jsha replied:
For the non-empty case we can just check that it's non-empty.

ccppuu replied:
Good idea. I was overthinking that we would need to match a specific predictable ID. A non-empty check would satisfy me.

jsha commented on wfe/wfe.go:
```diff
@@ -137,6 +139,8 @@ func (wfe *WebFrontEndImpl) HandleFunc(mux *http.ServeMux, pattern string, h wfe
 		log: wfe.log,
 		clk: clock.Default(),
 		wfe: wfeHandlerFunc(func(ctx context.Context, logEvent *requestEvent, response http.ResponseWriter, request *http.Request) {
+			response.Header().Set("Boulder-Request-ID", logEvent.ID)
```
This probably makes more sense in context.go right next to where the ID is created.

benileo commented:
Added not-empty check in test. 
Moved the addition of the Boulder-Requester header to `context.go`

Looks like CI failed for unrelated reason, as all tests passed locally.

jsha commented:
Yep, looks like CI hit a race in the CAA code that Roland is in the process of fixing. I restarted the build.

jsha commented on wfe/wfe.go:
```diff
@@ -537,6 +543,7 @@ func link(url, relation string) string {
 func (wfe *WebFrontEndImpl) NewRegistration(ctx context.Context, logEvent *requestEvent, response http.ResponseWriter, request *http.Request) {
 
 	body, key, _, prob := wfe.verifyPOST(ctx, logEvent, request, false, core.ResourceNewReg)
+	addRequesterHeader(response, logEvent.Requester)
```
This should come after the `prob != nil` check, and a similar change is needed below.

benileo replied:
I put it before `prob != nil` because there are a handful of errors that could be set in `verifyPOST` where `logEvent.Requester` would be set. 
If `addRequesterHeader` is after `prob != nil` responses would not contain the Requester header for responses that could potentially have the header set. If this is the desired functionality (or is a security issue) then I will make the changes as requested.

jsha replied:
Nope, you are right, and I had missed the check for requester>0 in addRequesterHeader.

jsha commented:
LGTM

ccppuu commented:
LGTM :dancers:
